src/TheThingsNetwork.cpp

```diff
@@ -219,29 +219,61 @@ bool TheThingsNetwork::personalize(const byte devAddr[4], const byte nwkSKey[16]
   return true;
 }
 
-bool TheThingsNetwork::join(const byte appEui[8], const byte appKey[16]) {
-  String devEui = readValue(F("sys get hweui"));
-  sendCommand(F("mac set appeui"), appEui, 8);
-  String str = "";
-  str.concat(F("mac set deveui "));
-  str.concat(devEui);
-  sendCommand(str);
-  sendCommand(F("mac set appkey"), appKey, 16);
-  if (!sendCommand(F("mac join otaa"))) {
-    debugPrintLn(F("Send join command failed"));
-    return false;
-  }
+bool TheThingsNetwork::join(long int nbr_delay, int max_attempts) {
+  int nbr_attempts = 0;
 
-  String response = readLine(10000);
-  if (response != F("accepted")) {
-    debugPrint(F("Join not accepted: "));
-    debugPrintLn(response);
-    return false;
+  while (nbr_attempts++ != max_attempts) {
+    delay(nbr_delay);
+    String devEui = readValue(F("sys get hweui"));
+    String str = "";
+    str.concat(F("mac set deveui "));
+    str.concat(devEui);
+    sendCommand(str);
+    if (!sendCommand(F("mac join otaa"))) {
+      debugPrintLn(F("Send join command failed"));
+      continue;
+    }
+    String response = readLine(10000);
+    if (response != F("accepted")) {
+      debugPrint(F("Join not accepted: "));
+      debugPrintLn(response);
+      continue;
+    }
+    debugPrint(F("Join accepted. Status: "));
+    debugPrintLn(readValue(F("mac get status")));
+    return true;
   }
+  return false;
+}
 
-  debugPrint(F("Join accepted. Status: "));
-  debugPrintLn(readValue(F("mac get status")));
-  return true;
+bool TheThingsNetwork::join(const byte appEui[8], const byte appKey[16], long int nbr_delay, int max_attempts) {
+  int nbr_attempts = 0;
+
+  reset();
+  while (nbr_attempts++ != max_attempts) {
```

Comment:
I think you could call `join()` (no args) here if you'd move out `mac set appeui` and `mac set appkey` out of the while. I'm not sure (@johanstokking?) but I think you don't need to do this before every attempt to `mac join otaa`?